Add NATS support for FHIR notifications (#973)
IBM/FHIR: docs/src/pages/guides/FHIRServerUsersGuide.md

On the other hand, if you have two completely independent FHIR server instances, then you should configure each one with its own topic name.

### 4.2.3 NATS

> The [NATS](http://nats.io) implementation of the notification service will publish notification event messages to a NATS streaming cluster. To configure the NATS notification publisher, configure properties in the `fhir-server-config.json` file as indicated in the following example:

Reviewer: ```suggestion
The [NATS](http://nats.io) implementation of the notification service publishes notification event messages to a NATS streaming cluster. To configure the NATS notification publisher, configure properties in the `fhir-server-config.json` file as shown in the following example:
```

```
{
    "fhirServer":{

>         …

Reviewer: ```suggestion
        ...
```

        "notifications":{

>             …

Reviewer: ```suggestion
            ...
```

            "nats": {
		        "enabled": true,
		        "cluster": "nats-streaming",
		        "channel": "fhirNotifications",
		        "clientId": "fhir-server",
		        "servers": "nats://nats-node1:4222,nats://nats-node2:4223,nats://nats-node3:4224",

> 		        "useTLS": false,

Reviewer: ```suggestion
		        "useTLS": true,
```
I suggest giving the TLS example.

		        "truststoreLocation": "resources/security/nats.client.truststore.p12",
		        "truststorePassword": "change-password",
		        "keystoreLocation": "resources/security/nats.client.keystore.p12",
		        "keystorePassword": "change-password"
    }

>         …

Reviewer: ```suggestion
        ...
```

Reviewer: Ellipsis are hard to see

    }
}
```

> Set the `fhirServer/notifications/nats/enabled` property to true and provide the name of your NATS cluster for the value of `fhirServer/notifications/nats/cluster`.  You may leave `fhirServer/notifications/nats/channel` and `fhirServer/notifications/nats/clientId` as defined.  Provide the URL for one or more NATS servers as the value for `fhirServer/notifications/nats/servers`. 

Reviewer: ```suggestion

Set the `fhirServer/notifications/nats/enabled` property to true and provide the name of your NATS cluster for the value of `fhirServer/notifications/nats/cluster`.  You may leave `fhirServer/notifications/nats/channel` and `fhirServer/notifications/nats/clientId` as defined.  Provide the URL for one or more NATS servers as the value for `fhirServer/notifications/nats/servers`. 
```

To use TLS to connect to your NATS cluster, set `fhirServer/notifications/nats/useTLS` to true and provide client truststore and keystore locations and passwords as the remaining config values. Ensure that your NATS cluster is configured for TLS client connections.

> To store a value requiring security (such as a password), you can use Liberty's securityUtility command to encode the value. See Section 3.1 Encoded passwords for details.

Reviewer: ```suggestion
To store a value requiring security, such as a password, use Liberty's `securityUtility` command to encode the value. See Section 3.1 Encoded passwords for details.
```

|`fhirServer/notifications/kafka/enabled`|boolean|A boolean flag which indicates whether or not kafka notifications are enabled.|
|`fhirServer/notifications/kafka/topicName`|string|The name of the topic to which kafka notification event messages should be published.|
|`fhirServer/notifications/kafka/connectionProperties`|property list|A group of connection properties used to configure the KafkaProducer. These properties are used as-is when instantiating the KafkaProducer used by the FHIR server for publishing notification event messages.|
|`fhirServer/notifications/nats/enabled`|boolean|A boolean flag which indicates whether or not NATS notifications are enabled.|
|`fhirServer/notifications/nats/cluster`|string|The name of the NATS streaming cluster to which to connect.|

> |`fhirServer/notifications/nats/channel`|string|The name of the NATS channel on which NATS notification event messages are to be published.|

Reviewer: worthy mentioning that there is only 1 channel for all tenants? 
@lmsurpre

|`fhirServer/notifications/nats/clientId`|string|The name to use for the connections to the NATS streaming cluster.|
|`fhirServer/notifications/nats/servers`|string|The URL of one or more NATS servers in the NATS streaming cluster.|
|`fhirServer/notifications/nats/useTLS`|boolean|A boolean flag which indicates whether or not to use TLS for connections to the NATS streaming cluster.|
|`fhirServer/notifications/nats/truststoreLocation`|string|The file location of the truststore to use for TLS.|

> |`fhirServer/notifications/nats/truststorePassword`|string|The password for the  truststore.|

Reviewer: ```suggestion
|`fhirServer/notifications/nats/truststorePassword`|string|The password for the truststore.|
```

|`fhirServer/bulkdata/applicationName`| string|Fixed value, always set to fhir-bulkimportexport-webapp |
|`fhirServer/bulkdata/moduleName`|string| Fixed value, always set to fhir-bulkimportexport.war |
|`fhirServer/bulkdata/jobParameters/cos.bucket.name`|string|Object store bucket name |
|`fhirServer/bulkdata/jobParameters/cos.location`|string|Object store location |
|`fhirServer/bulkdata/jobParameters/cos.endpointurl`|string|Object store end point url |

> |`fhirServer/bulkdata/jobParameters/credential.ibm`|string|If use IBM credential, "Y" or "N" |

Reviewer 2: good catch!

|`fhirServer/notifications/kafka/enabled`|false|
|`fhirServer/notifications/kafka/topicName`|fhirNotifications|
|`fhirServer/notifications/kafka/connectionProperties`|`{}`|
|`fhirServer/notifications/nats/enabled`|false|
|`fhirServer/notifications/nats/cluster`|nats-streaming|
|`fhirServer/notifications/nats/channel`|fhirNotifications|
|`fhirServer/notifications/nats/clientId`|fhir-server|
|`fhirServer/notifications/nats/servers`|nats://nats-node1:4222,nats://nats-node2:4223,nats://nats-node3:4224|

> |`fhirServer/notifications/nats/useTLS`|false|

Reviewer: just a thought to switch to true
it's a preferred security posture

Reviewer 2: i kinda gave it a pass since its new/experimental, but i suppose you're right.  the default should be secure.

Author: yes, except it's considerably easier to not use TLS if someone wants to try it out.  NATS cluster TLS can be time-consuming to configure properly.

|`fhirServer/notifications/kafka/enabled`|Y|Y|
|`fhirServer/notifications/kafka/topicName`|N|N|
|`fhirServer/notifications/kafka/connectionProperties`|N|N|
|`fhirServer/notifications/nats/enabled`|N|N|
|`fhirServer/notifications/nats/cluster`|Y|N|
|`fhirServer/notifications/nats/channel`|N|N|
|`fhirServer/notifications/nats/clientId`|N|N|
|`fhirServer/notifications/nats/servers`|Y|N|
|`fhirServer/notifications/nats/useTLS`|N|N|
|`fhirServer/notifications/nats/truststoreLocation`|Y|N|
|`fhirServer/notifications/nats/truststorePassword`|Y|N|
|`fhirServer/notifications/nats/keystoreLocation`|Y|N|

> |`fhirServer/notifications/nats/keystorePassword`|Y|N|

Reviewer: have we tested that this is tenant specific? 
how is this valid given the fact the channel is not tenant?

Author: Actually the channel could be tenant-specific, since it is configurable.  Probably worth a discussion about how the notification system would behave in a multi-tenant environment.  If there is fhir-server-config.json and nats notification class instance per tenant, then channel and client-id could be tenant-specific.  I was viewing that column from the perspective of whether a single tenant would need to change that value to something specific for that tenant, which is where the 'N's came from.

Reviewer 2: what we mean by "tenant-specific" here is whether the properties will be read from the tenant-specific config file if they are set there.  In this case, the nats notifier reads these properties during server startup (FHIRServletContextListener.contextInitialized), so we want an N in this column.  I've already committed the change, so this one is now resolved.
